### PR TITLE
Redux Architecture

### PR DESCRIPTION
**Pull request by arthur-here**
Yo, my lovely team, please, take your time reviewing this chapter. Feel free to ask any questions and give your feedback

### CODE REVIEW

**Comment by romanfurman6** [comment]
@arthur-here & @EvgenyMatviyenko Thank you guys, it looks amazing (especially Middleware). I really would like to try it on my current project. <3

**Comment by EvgenyMatviyenko** [comment]
@arthur-here I like the idea of applying Store and Middleware concepts to our way of doing things. However, it isn't clear for me now how middlewares should emit network errors to show, when we are not storing them in state. This applies not only to errors, but to every other hot observable triggers, like navigation to other screens etc
Beside that, everything else is clear and easy to follow. 
Thank you for great work done!

File: 4-1-redux.md [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+  case deleteMessage(id: Message.Identifier)
+}
+```
+
+## Reducer
+
+Reducer is a _pure_ function that applies actions to the state. It is important to keep it pure and implement it without side effects to be sure that the only way of modifying state is dispatching actions.
+
+```swift
+extension TeamChat {
```
[the remark this answers is not in the document]
Comment (arthur-here) [reply 4-1-redux.md]: there is no `TeamChat` object, I use it as a namespace or as a module. For example:
```swift
enum TeamChat { }

extension TeamChat {
  struct State {
    ...
  }
}

extension TeamChat {
  enum Action {
    ...
  }
}

extension TeamChat {
  typealias Store = ReduxStore<State, Action>

  static func reduce(state: State, action: Action) -> State { ... }
}

...
```

@tungfam do you think there is a way to make it less confusing? For example we can remove `extension TeamChat` completely

**Comment by arthur-here** [comment]
@EvgenyMatviyenko that's a very good question!

I believe that errors are part of the View state and should be handled like any other view property to keep stuff in sync. For example, what will happen if two errors are thrown at the same time? Let the reducer decide.

But there are cases when we **need** triggers, e.g. a navigation. It can be done pretty easily with a middlewares: 
```swift
private func makeRoutingMiddleware() -> (Store.Middleware, Signal<Route, NoError>) {
  let routeSignal = Signal<Route, NoError>.pipe()
  let middleware = Store.makeMiddleware { _, _, next, action in
    next(action)

    switch action {
    case .goNext:
      routeSignal.input.send(value: Route.next)
    }
  }

  return (middleware, routeSignal.output)
}

...

func makeOutputs(from inputs: Inputs) -> Outputs {
  let (routingMiddleware, routeSignal) = makeRoutingMiddleware()
  let store = Store(initialState: initialState, reducer: reduce, middlewares: [routingMiddleware])
  ...
  return Outpus(
    props: props,
    route: routeSignal
  )
}
```

Do you think we should add this example to the chapter?

**Comment by EvgenyMatviyenko** [comment]
@arthur-here That would be awesome if reducer decide what should we show to user, but that requires making functional alert controllers, that can be consistent with state. `UIAlertController` won't work this way, because it has imperative behavior of presenting and dismissing.
Yes, please add this example to the chapter. 
Thank you for your answer 😺

File: 4-1-redux.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (tungfam) [reply 4-1-redux.md]: oh i see. clear now.

but yeah maybe mention shortly what it is just to avoid a question like mine.

**Pull request merged**